### ISSUE TITLE
IPFS Feedback

### ISSUE THREAD

**Issue by jbenet** [ipfs/notes#318]
Hey all, would be great to get some feedback of the things you'd like IPFS to do better _right now_. This includes the go-ipfs repo, the ipfs protocol itself, the js webui, all our codebases, our community channels-- really, anything. There's of course **a lot** well known and planned-- i'm merely prioritizing where we spend our efforts in the coming weeks.

Some questions to get you thinking:
- I'd love it if it were easier to ....
- I wish that ...
- I really hate when ...
- It is very cumbersome to ...
- ... is very annoying
- ... is very broken

Thanks,
Juan

PS: yep, a rough spec is incoming shortly for those itching to implement in other langs. feel free to signal your desire here.

Crosspost from https://github.com/jbenet/go-ipfs/issues/849

**Comment by momack2** [comment]
Thanks for the feedback! @alanshaw - anything in the works on error handling in js-ipfs?

For the other requests - the go-ipfs gc fixes are actually in the upcoming 4.19 release, and we also have work underway on better provider strategies to improve content routing (follow along here: https://github.com/ipfs/go-ipfs/issues/5774)

**Comment by alanshaw** [comment]
> js-ipfs: Better error handling - i.e. getting some indication of when createReadStream is never going to return.

@mitra42 assuming you mean `ipfs.catReadableStream`? If it doesn't return, it means it's waiting for the content to be available. There's no "error" to handle but there's a few things we could do to help that aren't possible right now in JS IPFS:

1. Cancel the request - allow the request to be canceled by the user or otherwise
1. Timeout - if we don't find the content after 30s we could give up
1. Feedback on progress - provide an API whereby IPFS/libp2p can provide information about what it's doing for a given request e.g. checking local repo...checking connected peers...checking DHT. This would give users more information about what's happening and help them understand _why_ things are taking a long time and reassure them that IPFS hasn't just crashed!

We're working on it! https://github.com/ipfs/interface-ipfs-core/issues/58

Cancelling the request is the first step. Timeout is just cancel after a certain period of time.

I've been playing with [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) as a means of allowing this and I've not had any negative feedback yet 🤞. There's a couple of reasons for this - it's a native API in the browser and so should be familiar to frontend developers (it is used by `fetch`). I mentioned it on that same issue https://github.com/ipfs/interface-ipfs-core/issues/58#issuecomment-437809858

If we can pass an `AbortController.signal` to cancel a request we have a template for doing something similar with a per-request log to provide feedback on the status of any given request.

I aim to introduce this concept as we tackle the [async await/iterators endeavour](https://github.com/ipfs/js-ipfs/issues/1670).

**Comment by mitra42** [comment]
@alanshaw - I guess the thinking here is like BitTorrent, the user will wait as long as needed until someone comes online with the content.  But that's not how most applications work. For example, I want to open a file and load an image, the questions is, at what point does IPFS give up and tell the application so it can get the content somewhere else (e.g. via HTTP) - this is how WebTorrent works and is why its streaming experience is (from our perspective) so much better than IPFS's. 

I'm not sure how to do this well in IPFS, I can (and do) do a timeout, but this means that either 
a) the timeout is so long that it creates a painful user experience
b) the timeout is short enough that any large file will fail. 
I think - but don't know IPFS internals enough to be sure - that it wants a timeout that applies to making progress ... if IPFS is finding blocks, or getting closer in the DHT, then its fine to wait, but if progress is not being made then it needs to notify the app.

Note ... its the combination of IPFS having both low reliability (material stored in IPFS may or may not be retrievable at some later date, with some different connectivity); no errors; and no fallback to reliable methods is what makes IPFS performance appear so poor when its used as the primary way of for example loading content in a website.

**Comment by alanshaw** [comment]
Yes I agree the timeout should be tied to inactivity.

**Comment by mitra42** [comment]
Which means I think that  timeout is not as you said: `just a cancel after a period of time` because the timer needs to eb started/restarted in the js-ipfs layer, not the App.

**Comment by mitra42** [comment]
@alanshaw - I guess the thinking here is like BitTorrent, the user will wait as long as needed until someone comes online with the content.  But that's not how most applications work. For example, I want to open a file and load an image, the questions is, at what point does IPFS give up and tell the application so it can get the content somewhere else (e.g. via HTTP) - this is how WebTorrent works and is why its streaming experience is (from our perspective) so much better than IPFS's. 

I'm not sure how to do this well in IPFS, I can (and do) do a timeout, but this means that either 
a) the timeout is so long that it creates a painful user experience
b) the timeout is short enough that any large file will fail. 
I think - but don't know IPFS internals enough to be sure - that it wants a timeout that applies to making progress ... if IPFS is finding blocks, or getting closer in the DHT, then its fine to wait, but if progress is not being made then it needs to notify the app.

Note ... its the combination of IPFS having both low reliability (material stored in IPFS may or may not be retrievable at some later date, with some different connectivity); no errors; and no fallback to reliable methods is what makes IPFS performance appear so poor when its used as the primary way of for example loading content in a website.

**Comment by mitra42** [comment]
Did this ever get solved - so queries can give some kind of feedback/error on failure?

**Comment by Stebalien** [comment]
No but the issue is here: https://github.com/ipfs/go-ipfs/issues/5541. It's actually not all that difficult to solve, we just need a special `Reader` adapter that detects a stalled read. Curl actually has the same logic (it'll timeout after some time saying "no progress for N seconds".

**Comment by mitra42** [comment]
Pity its taken so long, I've run into a couple of groups this year who told me this was one of the key reasons they chose not to use IPFS.

**Comment by Stebalien** [comment]
Patches welcome, we're just limited on time.

**Comment by mitra42** [comment]
There's no way someone outside the IPFS core group could work on something like this - given that some issues relating to it have been open since 2016.  Its why in the dweb.archive.org site, we have IPFS as the last priority fallback (after http) in dweb.archive.org, because we can never tell if it is succeeding or failing we can't use IPFS and fallback to http, I always ask people what decentralized technologies they are using, and why and two projects told me it was the primary reason they dropped IPFS - I believe both of them were js-ipfs (you pointed at a go-ipfs issue). 

 In js-ipfs the topic gets pointed at the old (2016) issue https://github.com/ipfs/interface-ipfs-core/issues/58 which is about "cancellable requests" which  (without knowing the code) seems to be something different entirely.

**Comment by Stebalien** [comment]
An external contributor can, absolutely, fix that bug. There's just no way the core team can scale to meet the needs of every single user.

**Comment by whyrusleeping** [comment]
Also, since this is really just wanting a timeout for a javascript promise, you can implement it yourself pretty easily by taking the promise from js-ipfs, and racing it with a timeout: https://italonascimento.github.io/applying-a-timeout-to-your-promises/

**Comment by mitra42** [comment]
Yep - that timeout solution is the obvious one, the first thing we tried (almost a year ago) and unfortunately wrong :-( 
The problem is that failure is very different for a small or a large file, if the timeout is reasonable when for example IPFS is downloading the icons, (maybe 1 second) then its going to fail on larger files.  Its also wrong for streams, since streams return success immediately, failure is when they are unable to deliver data.

**Comment by OR13** [comment]
Would really love to be able to cancel stalled reads / shutdown the ipfs http client correctly. 

I use the promise timeout, but it does not help in jest.

Currently skipping unit tests that contain bad data, because there is no way to close ipfs properly once it starts trying to get a bad hash.
